File: frontend/src/services/authService.ts
import { get, post } from "./api";
import type { LoginRequest, LoginResponse, Utilisateur } from "../types";

interface MeResponse {
  utilisateur: Utilisateur;
  roles?: { id: number; nom: string; permissions: string[] }[];
}

export const authService = {
  async login(data: LoginRequest): Promise<LoginResponse> {
    const raw = await post<{ token: string; utilisateur: Utilisateur; roles?: unknown[] }>("/login", data);
    return { token: raw.token, utilisateur: raw.utilisateur };
  },

  logout(): Promise<void> {
    return post<void>("/logout");
  },

  async me(): Promise<Utilisateur> {
    const raw = await get<{ utilisateur: Utilisateur; personne?: Record<string, unknown> }>("/me");
    const user = raw.utilisateur;
    if (raw.personne && user) {
      (user as unknown as Record<string, unknown>).personne = raw.personne;
    }
    return user;
  },

  changePassword(data: { current_password: string; new_password: string; new_password_confirmation: string }): Promise<{ message: string }> {
    return post<{ message: string }>("/change-password", data);
  },
};

File: gs_operateur_mobile/lib/services/auth_service.dart
import '../models/user.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _api = ApiService();

  Future<Map<String, dynamic>> login(String email, String password) async {
    final response = await _api.post('/login', data: {
      'email': email,
      'password': password,
    });

    if (response['success'] == true) {
      final token = response['data']['token'];
      final user = User.fromJson(response['data']['utilisateur']);
      await _api.setToken(token);
      return {'token': token, 'user': user};
    }
    throw Exception(response['message'] ?? 'Erreur de connexion');
  }

  Future<User> me() async {
    final response = await _api.get('/me');
    if (response['success'] == true) {
      final data = response['data'];
      final userData = data['utilisateur'] ?? data;
      final rolesData = data['roles'] ?? userData['roles'];
      userData['roles'] = rolesData;
      return User.fromJson(userData);
    }
    throw Exception(response['message'] ?? 'Erreur lors du chargement du profil');
  }

  Future<void> logout() async {
    try {
      await _api.post('/logout');
    } finally {
      await _api.clearToken();
    }
  }

  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
    required String newPasswordConfirmation,
  }) async {
    final response = await _api.post('/change-password', data: {
      'current_password': currentPassword,
      'new_password': newPassword,
      'new_password_confirmation': newPasswordConfirmation,
    });
    if (response['success'] != true) {
      throw Exception(response['message'] ?? 'Erreur lors du changement de mot de passe');
    }
  }
}
